src/backend/drm/surface/mod.rs

```diff
@@ -391,125 +386,7 @@ impl DrmSurface {
-                        source,
-                    })?;
-                    // be careful here, we have no idea about the alignment inside the blob, so always copy using `read_unaligned`,
-                    // although slice::from_raw_parts would be so much nicer to iterate and to read.
-                    unsafe {
-                        let fmt_mod_blob_ptr = data.as_ptr() as *const drm_ffi::drm_format_modifier_blob;
-                        let fmt_mod_blob = &*fmt_mod_blob_ptr;
-
-                        let formats_ptr: *const u32 = fmt_mod_blob_ptr
-                            .cast::<u8>()
-                            .offset(fmt_mod_blob.formats_offset as isize)
-                            as *const _;
-                        let modifiers_ptr: *const drm_ffi::drm_format_modifier = fmt_mod_blob_ptr
-                            .cast::<u8>()
-                            .offset(fmt_mod_blob.modifiers_offset as isize)
-                            as *const _;
-                        let formats_ptr = formats_ptr as *const u32;
-                        let modifiers_ptr = modifiers_ptr as *const drm_ffi::drm_format_modifier;
-
-                        for i in 0..fmt_mod_blob.count_modifiers {
-                            let mod_info = modifiers_ptr.offset(i as isize).read_unaligned();
-                            for j in 0..64 {
-                                if mod_info.formats & (1u64 << j) != 0 {
-                                    let code = Fourcc::try_from(
-                                        formats_ptr
-                                            .offset((j + mod_info.offset) as isize)
-                                            .read_unaligned(),
-                                    )
-                                    .ok();
-                                    let modifier = Modifier::from(mod_info.modifier);
-                                    if let Some(code) = code {
-                                        formats.insert(Format { code, modifier });
-                                    }
-                                }
-                            }
-                        }
-                    }
-                }
-            }
-        } else if plane_type(self, plane)? == PlaneType::Cursor {
-            // Force a LINEAR layout for the cursor if the driver doesn't support modifiers
-            for format in formats.clone() {
-                formats.insert(Format {
-                    code: format.code,
-                    modifier: Modifier::Linear,
-                });
-            }
-        }
-
-        if formats.is_empty() {
-            formats.insert(Format {
-                code: Fourcc::Argb8888,
-                modifier: Modifier::Invalid,
-            });
-        }
-
-        trace!(
-            "Supported scan-out formats for plane ({:?}): {:?}",
-            plane,
-            formats
-        );
-
-        Ok(formats)
+        plane_formats(self, plane)
```

Comment:
Isn't it a bit weird, that we do enumerate this for `PlaneInfo`, but the `DrmSurface` still have to re-query this every time this function is called?

Reply:
`PlaneInfo` uses the same function under the hood, but neither are cached in the `DrmSurface`.  We can get rid of the `DrmSurface::supported_formats` call in the `DrmCompositor`, but we might want to keep the function for `GbmBufferedSurface`. Do you suggest to remove the `supported_formats` completely? Or to cache the PlaneInfo in the `DrmSurface` and return the formats from the cached info?

Reply:
Ah, we already call `planes` in `DrmDevice::create_surface`, so we could just put that once in the `DrmSurface`?

Reply:
> Ah, we already call planes in DrmDevice::create_surface, so we could just put that once in the DrmSurface?

Right, that is what I was thinking.

Reply:
Just pushed a commit with these changes, if you agree with them I would incorporate them correctly in the history and re-base the branch.

Reply:
Yep, looks good to me.